src/Symfony/Component/Security/Csrf/DoubleSubmitCsrfTokenManager.php

```diff
@@ -0,0 +1,201 @@
+ * send the X-Forwarded-* / Forwarded headers if you're behind one.
+ *
+ * Then, we validate the request using a cookie and a CsrfToken. If the cookie is found, it should
+ * contain the same value as the CsrfToken. A JavaScript snippet on the client side is responsible
+ * performing this double-submission. The token value should be regenerated on every request using
+ * a cryptographically secure random generator.
+ *
+ * If either double-submit or Origin/Referer headers are missing, it typically indicates that
+ * JavaScript is disabled on the client side, or that the JavaScript snippet was not properly
+ * implemented, or that the Origin/Referer headers were filtered out.
+ *
+ * Requests lacking both double-submit and origin information are deemed insecure.
+ *
+ * When a session is found, a behavioral check is added to ensure that the validation method does not
+ * downgrade from double-submit to origin checks. This prevents attackers from exploiting potentially
+ * less secure validation methods once a more secure method has been confirmed as functional.
+ *
+ * On HTTPS connections, the cookie is prefixed with "__Host-" to prevent it from being forged on an
+ * HTTP channel. On the JS side, the cookie should be set with samesite=strict to strengthen the CSRF
+ * protection. The cookie is always cleared on the response to prevent any further use of the token.
+ *
+ * In addition to a cookie, the token is also looked for in a header when the $checkHeader argument
+ * is true. This makes it even harder for an attacker to forge a request, but it could also be more
+ * difficult to set this header in JavaScript depending on the client-side framework in use.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class DoubleSubmitCsrfTokenManager implements CsrfTokenManagerInterface
+{
+    public const TOKEN_MIN_LENGTH = 24;
+
+    public function __construct(
+        private RequestStack $requestStack,
+        private ?LoggerInterface $logger = null,
+        private bool $checkHeader = false,
+        private string $cookieName = 'x-csrf-token',
+    ) {
+    }
+
+    public function getToken(string $tokenId): CsrfToken
+    {
+        return new CsrfToken($tokenId, $this->cookieName);
+    }
+
+    public function refreshToken(string $tokenId): CsrfToken
+    {
+        return new CsrfToken($tokenId, $this->cookieName);
+    }
+
+    public function removeToken(string $tokenId): ?string
+    {
+        return null;
+    }
+
+    public function isTokenValid(CsrfToken $token): bool
+    {
+        if (!$request = $this->requestStack->getCurrentRequest()) {
+            $this->logger?->debug('CSRF validation failed: No request found.');
+
+            return false;
+        }
+
+        if (false === $isValidOrigin = $this->isValidOrigin($request)) {
+            $this->logger?->debug('CSRF validation failed: origin info doesn\'t match.');
```

Comment:
Which level should we give to failures? It might be important to monitor them.

Reply:
INFO at least IMHO